executor/split.go

```diff
@@ -75,10 +79,128 @@ func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.RecordBatch) e
+	var idxKeys [][]byte
+	if e.num > 0 {
+		idxKeys = make([][]byte, 0, e.num)
+	} else {
+		idxKeys = make([][]byte, 0, len(e.valueLists)+1)
+	}
+	// Split in the start of the index key.
+	startIdxKey := tablecodec.EncodeTableIndexPrefix(e.tableInfo.ID, e.indexInfo.ID)
+	idxKeys = append(idxKeys, startIdxKey)
+
+	index := tables.NewIndex(e.tableInfo.ID, e.tableInfo, e.indexInfo)
+	// Split index regions by user specified value lists.
+	if len(e.valueLists) > 0 {
+		for _, v := range e.valueLists {
+			idxKey, _, err := index.GenIndexKey(e.ctx.GetSessionVars().StmtCtx, v, math.MinInt64, nil)
+			if err != nil {
+				return nil, err
+			}
+			idxKeys = append(idxKeys, idxKey)
+		}
+		return idxKeys, nil
+	}
+	// Split index regions by lower, upper value and calculate the step by (upper - lower)/num.
+	lowerIdxKey, _, err := index.GenIndexKey(e.ctx.GetSessionVars().StmtCtx, e.lower, math.MinInt64, nil)
+	if err != nil {
+		return nil, err
+	}
+	upperIdxKey, _, err := index.GenIndexKey(e.ctx.GetSessionVars().StmtCtx, e.upper, math.MinInt64, nil)
+	if err != nil {
+		return nil, err
+	}
+	if bytes.Compare(lowerIdxKey, upperIdxKey) >= 0 {
+		lowerStr, err1 := datumSliceToString(e.lower)
+		upperStr, err2 := datumSliceToString(e.upper)
+		if err1 != nil || err2 != nil {
+			return nil, errors.Errorf("Split index region `%v` lower value %v should less than the upper value %v", e.indexInfo.Name, e.lower, e.upper)
+		}
+		return nil, errors.Errorf("Split index region `%v` lower value %v should less than the upper value %v", e.indexInfo.Name, lowerStr, upperStr)
+	}
+	return getValuesList(lowerIdxKey, upperIdxKey, e.num, idxKeys), nil
+}
+
+// getValuesList is used to get `num` values between lower and upper value.
+// To Simplify the explain, suppose lower and upper value type is int64, and lower=0, upper=100, num=10,
+// then calculate the step=(upper-lower)/num=10, then the function should return 0+10, 10+10, 20+10... all together 9 (num-1) values.
+// then the function will return [10,20,30,40,50,60,70,80,90].
+// The difference is the value type of upper,lower is []byte, So I use getUint64FromBytes to convert []byte to uint64.
+func getValuesList(lower, upper []byte, num int, valuesList [][]byte) [][]byte {
+	commonPrefixIdx := longestCommonPrefixLen(lower, upper)
+	step := getStepValue(lower[commonPrefixIdx:], upper[commonPrefixIdx:], num)
+	startV := getUint64FromBytes(lower[commonPrefixIdx:], 0)
+	// To get `num` regions, only need to split `num-1` idx keys.
+	buf := make([]byte, 8)
+	for i := 0; i < num-1; i++ {
+		value := make([]byte, 0, commonPrefixIdx+8)
+		value = append(value, lower[:commonPrefixIdx]...)
+		startV += step
+		binary.BigEndian.PutUint64(buf, startV)
+		value = append(value, buf...)
+		valuesList = append(valuesList, value)
+	}
+	return valuesList
+}
+
+// longestCommonPrefixLen gets the longest common prefix byte length.
+func longestCommonPrefixLen(s1, s2 []byte) int {
+	l := len(s1)
```

Comment:
It's so weird that there's no builtin `min` function for integers in Go.

Reply:
done. I have used mathutil pkg.